### PR TITLE
Improve the WooCommerce build process

### LINKED ISSUE #28717: Spike: Analyze the build process for potential optimization
Desired outcome: Proposal on how to best optimize the build process for WooCommerce core. 

Potential areas to examine: 
- text domain changes and pot file generation only for final builds (and artifacts)
- bootstrapping lerna each time
- managing dependencies
- e2e test build with Chromium only when you need it
- more granular builds
- concurrent processes in node
- a new -build script- package.json with options and stuff?
- WC Admin will be probably merged to our repo, so that might affect it quite a bit, e.g. usage of webpack

- possible tool: https://www.npmjs.com/package/@telus/build-essential

### PR DESCRIPTION
**Pull request by Konamiman**
### All Submissions:

* [ ] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [ ] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [ ] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

* Create an e2e-less build script.
  * Use [the `install-subset` package](https://www.npmjs.com/package/install-subset) to define a subset of the NPM dependencies that doesn't include the e2e testing infrastructure.
  * Define a new NPM `install:no-e2e` command taht makes use of `install-subset` to do an e2e-less install.
  * Modify the git post-merge hook to use the new command instead of the full install.
  * `npm install` still installs everything, including e2e testing infrastructure.
<br/>

* Skip updating textdomains on `composer install` when it's triggered by the git post-merge hook.
<br/>

* Remove duplicate processing from the zip build process.
  * The `build:zip` NPM command was executing `build-zip.sh` and then doing `composer install`, `grunt` and `makepot`; but these three are already done as part of `build-zip.sh`.
<br/> 

* Exclude the e2e infrastructure packages from the zip build process.
  * As part of this the `require( '@woocommerce/e2e-environment' );` from `.eslintrc.js` has been removed, it doesn't seem to be needed and it prevents the build process to finish without the e2e infrastructure.

Closes https://github.com/woocommerce/woocommerce/issues/28717.

### How to test the changes in this Pull Request:

1. Try the new `npm run install:no-e2e` command, verify it installs everything needed but the e2e infrastructure (and lerna doesn't run).
2. Run `SKIP_UPDATE_TEXTDOMAINS=true composer install`, veryfy that text domains are not updated but otherwise the rest of the install is successful.
3. Verify that `npm install` still includes the e2e infrastructure in the install, and that you are still able to run the e2e tests after [following the instructions](https://github.com/woocommerce/woocommerce/blob/master/tests/e2e/README.md).

### Other information:

* [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?

### Changelog entry

> Dev - Optimize the build process by de-duplicating processing and providing a new e2e-less NPM install command.

### CODE REVIEW

File: .eslintrc.js
Comment (rrennick) [inline .eslintrc.js:3]: @Konamiman Your PR description doesn't explain why you removed this. The reason for this function call is related to running E2E tests.

Comment (Konamiman) [reply .eslintrc.js:3]: It seemed to me that this was about litning the js only and it was not really needed, but I'll take a closer look, thanks.

Comment (Konamiman) [reply .eslintrc.js:3]: Ok, looks like that `require( '@woocommerce/e2e-environment' );` isn't actually needed to run the e2e tests, and having it in place prevents the e2e infrastructure to be removed from the build. So indeed it needs to go out, I'll added that explanation to the PR description.

Comment (rrennick) [reply .eslintrc.js:3]: Sorry, when I said running E2E tests I generalized that it was needed for E2E. The linter is the js equivalent of phpcs for writing E2E tests. Instead of removing this can you change the import to 

```
const { useE2EEsLintConfig } = require( './tests/e2e/env/config/use-config' );
```

That will allow you to eliminate the package but keep the existing linting process in place.

Comment (Konamiman) [reply .eslintrc.js:3]: Thanks for your suggestion, but by applying it I still get this when running `grunt`: `Error: Cannot find module '@woocommerce/e2e-environment'`

Comment (rrennick) [reply .eslintrc.js:3]: This log is based against the latest on this branch

```
$ git diff
diff --git a/.eslintrc.js b/.eslintrc.js
index 7c90a4dbb..c64f82197 100644
--- a/.eslintrc.js
+++ b/.eslintrc.js
@@ -1,6 +1,7 @@
 /** @format */
+const { useE2EEsLintConfig } = require( './tests/e2e/env/config/use-config' );
 
-module.exports = {
+module.exports = useE2EEsLintConfig( {
        root: true,
        env: {
                browser: true,
@@ -28,4 +29,4 @@ module.exports = {
                        jsx: true
                }
        },
-};
+} );
$ grunt
Running "eslint:src" (eslint) task
WARNING: No configurations found in configuration directory:/Users/ronrennick/Sites/solaris/woocommerce/config
WARNING: To disable this warning set SUPPRESS_NO_CONFIG_WARNING in the environment.

/Users/ronrennick/Sites/solaris/woocommerce/assets/js/admin/system-status.js
  76:5  warning  Unexpected console statement  no-console

/Users/ronrennick/Sites/solaris/woocommerce/assets/js/frontend/checkout.js
  493:9  warning  Unexpected console statement  no-console
  495:9  warning  Unexpected console statement  no-console
  496:9  warning  Unexpected console statement  no-console
  499:9  warning  Unexpected console statement  no-console
  670:7  warning  Unexpected console statement  no-console

✖ 6 problems (0 errors, 6 warnings)


Running "uglify:js_assets" (uglify) task
>> 70 files created 1.56 MB → 700.46 kB

Running "sass:compile" (sass) task

Running "rtlcss:woocommerce" (rtlcss) task
File assets/css/activation-rtl.css created.
File assets/css/admin-rtl.css created.
File assets/css/auth-rtl.css created.
File assets/css/dashboard-rtl.css created.
File assets/css/helper-rtl.css created.
File assets/css/marketplace-suggestions-rtl.css created.
File assets/css/menu-rtl.css created.
File assets/css/network-order-widget-rtl.css created.
File assets/css/prettyPhoto-rtl.css created.
File assets/css/privacy-rtl.css created.
File assets/css/reports-print-rtl.css created.
File assets/css/twenty-nineteen-rtl.css created.
File assets/css/twenty-seventeen-rtl.css created.
File assets/css/twenty-twenty-one-admin-rtl.css created.
File assets/css/twenty-twenty-one-rtl.css created.
File assets/css/twenty-twenty-rtl.css created.
File assets/css/wc-setup-rtl.css created.
File assets/css/woocommerce-layout-rtl.css created.
File assets/css/woocommerce-smallscreen-rtl.css created.
File assets/css/woocommerce-rtl.css created.

Running "postcss:dist" (postcss) task
>> 41 processed stylesheets created.

Running "cssmin:minify" (cssmin) task
>> 43 files created. 1.51 MB → 1.24 MB

Running "concat:admin" (concat) task

Done.
```

Comment (Konamiman) [reply .eslintrc.js]: I must have something misconfigured then. Anyway, I have pushed the change you suggested; could someone else from @woocommerce/proton test this please?

**Comment by rrennick** [comment]
@Konamiman It wasn't in the test instructions but I wanted to verify that the GH action build zip was still creating a valid/working plugin. 

The plugin is built correctly but (probably unrelated to this PR) the download file is `woocommerce.zip.zip` (a `zip` of of a `zip`). If people download the file and try to install it with the WP plugin installer it will fail. We should look at adjusting the GH action to not zip the artifact.

**Review by rrennick — APPROVED** [review]
@Konamiman Nice work on this. It's great to see these optimization PRs. I'm approving the PR but will leave it for @ObliviousHarmony for a second review before merging.

**Review by claudiosanches — APPROVED** [review]
Great work!

**Review by ObliviousHarmony — APPROVED** [review]
To preface I deleted all `node_modules` folders using `rm -rf node_modules tests/e2e/**/node_modules` before running anything and ensured `nvm use` was ran.

I received this after running the install without E2E content:

```
(node:55104) Warning: Accessing non-existent property 'cat' of module exports inside circular dependency
(Use `node --trace-warnings ...` to show where the warning was created)
(node:55104) Warning: Accessing non-existent property 'cd' of module exports inside circular dependency
(node:55104) Warning: Accessing non-existent property 'chmod' of module exports inside circular dependency
(node:55104) Warning: Accessing non-existent property 'cp' of module exports inside circular dependency
(node:55104) Warning: Accessing non-existent property 'dirs' of module exports inside circular dependency
(node:55104) Warning: Accessing non-existent property 'pushd' of module exports inside circular dependency
(node:55104) Warning: Accessing non-existent property 'popd' of module exports inside circular dependency
(node:55104) Warning: Accessing non-existent property 'echo' of module exports inside circular dependency
(node:55104) Warning: Accessing non-existent property 'tempdir' of module exports inside circular dependency
(node:55104) Warning: Accessing non-existent property 'pwd' of module exports inside circular dependency
(node:55104) Warning: Accessing non-existent property 'exec' of module exports inside circular dependency
(node:55104) Warning: Accessing non-existent property 'ls' of module exports inside circular dependency
(node:55104) Warning: Accessing non-existent property 'find' of module exports inside circular dependency
(node:55104) Warning: Accessing non-existent property 'grep' of module exports inside circular dependency
(node:55104) Warning: Accessing non-existent property 'head' of module exports inside circular dependency
(node:55104) Warning: Accessing non-existent property 'ln' of module exports inside circular dependency
(node:55104) Warning: Accessing non-existent property 'mkdir' of module exports inside circular dependency
(node:55104) Warning: Accessing non-existent property 'rm' of module exports inside circular dependency
(node:55104) Warning: Accessing non-existent property 'mv' of module exports inside circular dependency
(node:55104) Warning: Accessing non-existent property 'sed' of module exports inside circular dependency
(node:55104) Warning: Accessing non-existent property 'set' of module exports inside circular dependency
(node:55104) Warning: Accessing non-existent property 'sort' of module exports inside circular dependency
(node:55104) Warning: Accessing non-existent property 'tail' of module exports inside circular dependency
(node:55104) Warning: Accessing non-existent property 'test' of module exports inside circular dependency
(node:55104) Warning: Accessing non-existent property 'to' of module exports inside circular dependency
(node:55104) Warning: Accessing non-existent property 'toEnd' of module exports inside circular dependency
(node:55104) Warning: Accessing non-existent property 'touch' of module exports inside circular dependency
(node:55104) Warning: Accessing non-existent property 'uniq' of module exports inside circular dependency
(node:55104) Warning: Accessing non-existent property 'which' of module exports inside circular dependency
```

Everything did work as expected though so I'm not entirely sure what's up there.

I'll admit I have some concerns about this not being comprehensive enough but any improvement on our build pipeline is a step in the right direction so I'll reserve those for another sprint task!

**Pull request merged**
